library/core/src/primitive_docs.rs

```diff
@@ -1676,6 +1676,11 @@ mod prim_ref {}
 /// * [`UnwindSafe`]
 /// * [`RefUnwindSafe`]
 ///
+/// Note that while this type implements `PartialEq`, comparing function pointers is unreliable:
+/// pointers to the same function can compare inequal (because functions are duplicated in multiple
+/// codegen units), and pointers to *different* functions can compare equal (since identical
+/// functions can be deduplicated within a codegen unit).
+///
 /// [`Hash`]: hash::Hash
```

Comment:
I hadn't considered before -- are we then violating the `Hash`-`Eq` contract? Or is it ok because the separate origins of pointers to the "same" function will at least be self-consistent in hashing, and pointers to different deduplicated functions will also hash the same.

That is, if you use function pointers as a `HashMap` key, they will behave correctly with respect to the address of actual codegen functions, but not their original function in the source code.

Reply:
Hash and PartialEq use the same source data, so that contract is maintained.

> That is, if you use function pointers as a HashMap key, they will behave correctly with respect to the address of actual codegen functions, but not their original function in the source code.

Correct.